src/TorchSharp/Tensor/Tensor.Factories.cs

```diff
@@ -2045,6 +2045,87 @@ public static Tensor tensor(IList<System.Numerics.Complex> rawArray, long dim0,
+        /// <code>
+        /// var array = new double[,,] { { { 1, 2 }, { 3, 4 } }, { { 5, 6 }, { 7, 8 } } };
+        /// var tensor = torch.tensor_from_array(rawArray: array, dtype: torch.ScalarType.Float64, device: torch.Device.CPU, requiresGrad: false);
+        /// </code>
+        /// </example>
+        public static Tensor tensor_from_array(Array rawArray, ScalarType? dtype = null, Device? device = null, bool requiresGrad = false)
+        {
+            // enumerates over all dimensions of the arbitrary array
+            // and returns the length of the dimension
+            IEnumerable<long> GetShape(Array arr)
+            {
+                for (var dim = 0; dim < arr.Rank; dim++) {
+                    var dimLength = arr.GetLength(dim);
+                    yield return dimLength;
+                }
+            }
+            var shape = GetShape(rawArray).ToArray();
+
+            var t = rawArray.GetType().GetElementType();
+            if (t is null) throw new InvalidOperationException($"{nameof(rawArray)}.GetType().GetElementType() returned null.");
+
+            IEnumerable<object> GetElements(Array rawArray)
+            {
+                foreach (var element in rawArray) {
+                    yield return element!;
+                }
+            }
+            var elements = GetElements(rawArray);
+
+            // call the existing factory methods to construct the tensor
+            if (t == typeof(bool)) return tensor(elements.Cast<bool>().ToArray(), shape, dtype, device, requiresGrad);
+            if (t == typeof(byte)) return tensor(elements.Cast<byte>().ToArray(), shape, dtype, device, requiresGrad);
+            if (t == typeof(sbyte)) return tensor(elements.Cast<sbyte>().ToArray(), shape, dtype, device, requiresGrad);
+            if (t == typeof(short)) return tensor(elements.Cast<short>().ToArray(), shape, dtype, device, requiresGrad);
+            if (t == typeof(int)) return tensor(elements.Cast<int>().ToArray(), shape, dtype, device, requiresGrad);
+            if (t == typeof(long)) return tensor(elements.Cast<long>().ToArray(), shape, dtype, device, requiresGrad);
+#if NET50_OR_GREATER
+            // TODO: implement the required factory method
+            // if (t == typeof(half)) return tensor(elements.Cast<half>().ToArray(), shape, dtype, device, requiresGrad);
+#endif
+            if (t == typeof(float)) return tensor(elements.Cast<float>().ToArray(), shape, dtype, device, requiresGrad);
+            if (t == typeof(double)) return tensor(elements.Cast<double>().ToArray(), shape, dtype, device, requiresGrad);
+            // TODO: implement the required factory methods
+            // if (t == typeof(Tuple<float, float>)) return tensor(elements.Cast<Tuple<float, float>>().ToArray(), shape, dtype, device, requiresGrad);
+            // if (t == typeof(Tuple<double, double>)) return tensor(elements.Cast<Tuple<double, double>>().ToArray(), shape, dtype, device, requiresGrad);
```

Comment:
We don't use (double,double) to represent 64-bit complex, and the (float,float) factory methods should already exist.